### DISCUSSION TITLE
Mapper notes

### DISCUSSION THREAD

**Discussion by sakkammadam** [sakkammadam/scratchStuff#5] [Announcements]
Hi @hbaird7877, Abraham - 
 I wanted to let you know you I started on the Mapper class. I have completed tokenization and token count operations.
 Please look at the latest commits on the sa_branch.
 There are two new files - Mapper.h and Mapper.cpp. I have updated the main.cpp with test code to output the data for a single file.
 Please run it when you have a moment.
 I have integrated Boost but I skipped using the boost's string lower functionality. 
 
 I have NOT implemented the export functionality - I think the implementation should live within the FileProcessor class. We would simply intro. the FileProcessor class as an object within the export method.

Per the requirements - 
The export function will buffer output in memory and periodically write the data out to disk (periodicity can be based on the size of the buffer).

Could one of you research how to do this ? This would be greatly helpful. 

 Let me know if you have any questions/concerns. Feel free to text me on WhatApp or call me - I should be able to talk after 11:30 am PST, except for a 2 to 2:30 pm work meeting.

Thanks,
Sooraj

**Comment by hbaird7877** [comment]
sakkammadam stated, "Per the requirements -
The export function will buffer output in memory and periodically write the data out to disk (periodicity can be based on the size of the buffer).

Could one of you research how to do this ? This would be greatly helpful."

I am working on this buffered output stream issue.

Focusing on an output stream that buffers until a specified number of lines have been buffered and then writing out to disk.

Thanks, Hal

**Comment by sakkammadam** [comment]
Hi @hbaird7877 , @abejr17  - 
 Guys - I have updated the Mapper class to use tuples vs. map for the tokens. This reflects the partition line number. All data belonging to the same partition line number can be buffered to same output - we can use this as the periodicity.
 Please look at the changes I have made, if you want to analyze the new output - please look at mapOutput.dat

![explainMapper](https://user-images.githubusercontent.com/10201590/196349795-9c734a06-491b-4685-8701-f1aec6de2e60.jpg)

Notice the partition number. When this changes we need to write all data in the previous portion into the disk.

I hope to meet with you folks tomorrow.

Best,
Sooraj

**Comment by sakkammadam** [comment]
Team -
 I have some bad news, we have to revisit the mapper internals a bit.
 I know we are back tracking but this make the remaining implementation of FileProcessor a bit easier to implement.
 I couldn't finish it last night, I didn't want to start on a breaking change late at night. I will work on it today after class.
 I will explain during class what we are planning to implement.


Thanks,
Sooraj

**Comment by sakkammadam** [comment]
@hbaird7877, @abejr17 
Team - 
 I want to let you know that the mapper class implementation is complete.
 I have also updated the FileProcessor class to write out the results of the mapper operations to the file system.
 Please take a look at my commits and comments for further analysis. 
 If I may suggest, please create new branch from the latest sa_branch. Due to the change in the data structure, there is a significant difference in the code compared to previously. The newer implementation is a bit more straightforward than before.
 I have left areas in the code where we should comment, lets try to get the logging working tomorrow.

 I ran two tests - 1) Folder had a only a single file, 2) Folder had all Shakespeare files

The mapper output data will get written to a folder within the input folder called temp_mapper

![Screenshot from 2022-10-20 01-21-36](https://user-images.githubusercontent.com/10201590/196895728-660e5f89-5488-4dcc-b197-498acf171cd3.png)

Within temp_mapper folder, you will see sub-folders created for each file - 
![Screenshot from 2022-10-20 01-22-52](https://user-images.githubusercontent.com/10201590/196896042-22e60938-1d10-4fcf-bea9-22b643214090.png)

On clicking each sub-folder (file) - 
![Screenshot from 2022-10-20 01-23-21](https://user-images.githubusercontent.com/10201590/196896177-52920bfd-7772-4039-906a-319a0c9d51be.png)

You will see files being broken down to partitions. - 
![Screenshot from 2022-10-20 01-24-05](https://user-images.githubusercontent.com/10201590/196896424-f33616f2-ae83-4528-af62-1a9b3dd9064d.png)

Each partitioned file will have tokenized data.

Please let me know if you have any questions.

Thanks,
Sooraj

**Comment by sakkammadam** [comment]
@hbaird7877 / @abejr17 
Zoom recording from earlier - 
https://zoom.us/rec/play/ez-WjiTl_iUuSfA-6zsDajYLOa85r1BLhYTFLEJR2RBUx5tkIQKwaVnnz62PYwfvoWb-v2sO8xXC80-i.D0qa7lZcB8mhlILS

**Reply by hbaird7877** [reply]
I tried several times to watch the recording of last evening's Zoom recording regarding github from last night.  The system asked me to log in, which I did but it was to no avail. Perhaps one of you has an idea what I need to do to be able to view the Zoom recording.  Thanks, Hal

**Comment by sakkammadam** [comment]
@hbaird7877  - could you confirm if you have logged into your 2SU website prior to clicking the Zoom recording link ?

**Reply by hbaird7877** [reply]
Yes I was.  Tonight I specifically logged into 2SU before I tried to watch the recording.  Juat like the day before, the system asked me to log in again when I clicked on the recording.  Each time I logged in with my 2SU credentials.  Each time I was denied access. Thanks. Hal

**Reply by hbaird7877** [reply]
Each time the system reports, "You cannot view this recording.
No permission."
